src/Features/Blockcore.Features.BlockStore/Api/Controllers/BlockStoreRPCController.cs

```diff
@@ -0,0 +1,107 @@
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.Extensions.Logging;
+using NBitcoin;
+
+namespace Blockcore.Features.BlockStore.Api.Controllers
+{
+    /// <summary>
+    /// Controller providing RPC operations on a watch-only wallet.
+    /// </summary>
+    public class BlockStoreRPCController : FeatureController
+    {
+        /// <summary>Thread safe access to the best chain of block headers from genesis.</summary>
+        private readonly ChainIndexer chainIndexer;
+
+        /// <summary>Provides access to the block store database.</summary>
+        private readonly IBlockStore blockStore;
+
+        /// <inheritdoc />
+        public BlockStoreRPCController(
+            IFullNode fullNode,
+            IConsensusManager consensusManager,
+            ChainIndexer chainIndexer,
+            Network network,
+            IBlockStore blockStore) : base(fullNode: fullNode, consensusManager: consensusManager, chainIndexer: chainIndexer, network: network)
+        {
+            this.chainIndexer = chainIndexer;
+            this.blockStore = blockStore;
+        }
+
+        /// <summary>
+        /// By default this function only works when there is an unspent output in the utxo for this transaction.
+        /// To make it work, you need to maintain a transaction index, using the -txindex command line option.
+        /// </summary>
+        /// <param name="txids">The txids to filter</param>
+        /// <param name="blockhash">If specified, looks for txid in the block with this hash</param>
+        /// <returns></returns>
+        [ActionName("gettxoutproof")]
+        [ActionDescription("Checks if transactions are within block. Returns proof of transaction inclusion (raw MerkleBlock).")]
+        public MerkleBlock GetTxOutProof(string[] txids, string blockhash = "")
+        {
+            List<uint256> transactionIds = new List<uint256>();
+            Block block = null;
+            foreach (var txString in txids)
+            {
+                transactionIds.Add(uint256.Parse(txString));
+            }
+
+            if (!string.IsNullOrEmpty(blockhash))
+            {
+                // Loop through txids and veryify that the transaction is in the block.
+                foreach (var transactionId in transactionIds)
+                {
+                    var hashBlock = uint256.Parse(blockhash);
+                    ChainedHeader chainedHeader = this.GetTransactionBlock(transactionId);
+                    if (chainedHeader.HashBlock != hashBlock)
+                    {
+                        throw new RPCServerException(RPCErrorCode.RPC_INVALID_ADDRESS_OR_KEY, "Not all transactions found in specified or retrieved block");
+                    }
+                    if (block == null && chainedHeader.BlockDataAvailability == BlockDataAvailabilityState.BlockAvailable) // Only get the block once.
+                    {
+                        block = this.blockStore.GetBlock(chainedHeader.HashBlock);
```

Comment:
aren't you missing a `break;` here?
Do we assume all trx are of the same block?

Reply:
Ideally to fetch blocks we should use the ConsensusManager that will fetch from memory first (and it can even download a block if its not available.

Reply:
> aren't you missing a `break;` here?
> Do we assume all trx are of the same block?

No, we need to check all transactions.

Reply:
> Ideally to fetch blocks we should use the ConsensusManager that will fetch from memory first (and it can even download a block if its not available.

Done. Good to know, thank you.

Reply:
> No, we need to check all transactions.

but once you find the block the param is not null and wont be fetched anymore, unless you want to find all the headers and check they are part of the same block

Reply:
> > No, we need to check all transactions.
> 
> but once you find the block the param is not null and wont be fetched anymore, unless you want to find all the headers and check they are part of the same block

That is correct, I want to verify that they are all in the same block, otherwise, it's not valid proof.